### PR TITLE
Decision task handling refactor

### PR DESCRIPTION
**Pull request by wxing1292**
* Move all decsion attribute validation into decisionAttrValidator
* Move decisionBlobSizeChecker and decisionAttrValidator into one file
* New decisionTaskHandler which handles decision validation, size checking and processing.
* Refactor handleDecisionTaskCompleted by using decisionTaskHandler

### CODE REVIEW

File: service/history/decisionChecker.go [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+func (v *decisionAttrValidator) validateActivityScheduleAttributes(
+	attributes *workflow.ScheduleActivityTaskDecisionAttributes,
+	wfTimeout int32,
+) error {
+
+	if attributes == nil {
+		return &workflow.BadRequestError{Message: "ScheduleActivityTaskDecisionAttributes is not set on decision."}
+	}
+
+	if attributes.TaskList == nil || attributes.TaskList.Name == nil || *attributes.TaskList.Name == "" {
```
Comment (yux0) [inline service/history/decisionChecker.go]: attributes.TaskList.GetName()

File: service/history/decisionChecker.go [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+		return &workflow.BadRequestError{Message: "TaskList is not set on decision."}
+	}
+
+	if attributes.ActivityId == nil || *attributes.ActivityId == "" {
```
Comment (yux0) [inline service/history/decisionChecker.go]: attributes.GetActivitId()

File: service/history/decisionChecker.go [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+		return &workflow.BadRequestError{Message: "ActivityId is not set on decision."}
+	}
+
+	if attributes.ActivityType == nil || attributes.ActivityType.Name == nil || *attributes.ActivityType.Name == "" {
```
Comment (yux0) [inline service/history/decisionChecker.go]: attributes.GetActivityType().GetName()

File: service/history/decisionChecker.go [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+}
+
+func (v *decisionAttrValidator) validateTimerScheduleAttributes(
+	attributes *workflow.StartTimerDecisionAttributes,
+) error {
+
+	if attributes == nil {
+		return &workflow.BadRequestError{Message: "StartTimerDecisionAttributes is not set on decision."}
+	}
+	if attributes.TimerId == nil || *attributes.TimerId == "" {
```
Comment (yux0) [inline service/history/decisionChecker.go]: attributes.GetTimerId()

File: service/history/decisionChecker.go [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+		return &workflow.BadRequestError{Message: "TimerId is not set on decision."}
+	}
+	if len(attributes.GetTimerId()) > v.maxIDLengthLimit {
+		return &workflow.BadRequestError{Message: "TimerId exceeds length limit."}
+	}
+	if attributes.StartToFireTimeoutSeconds == nil || *attributes.StartToFireTimeoutSeconds <= 0 {
```
Comment (yux0) [inline service/history/decisionChecker.go]: attributes.GetStartToFireTimeoutSeconds()

File: service/history/decisionChecker.go [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+}
+
+func (v *decisionAttrValidator) validateActivityCancelAttributes(
+	attributes *workflow.RequestCancelActivityTaskDecisionAttributes,
+) error {
+
+	if attributes == nil {
+		return &workflow.BadRequestError{Message: "RequestCancelActivityTaskDecisionAttributes is not set on decision."}
+	}
+	if attributes.ActivityId == nil || *attributes.ActivityId == "" {
```
Comment (yux0) [inline service/history/decisionChecker.go]: attributes.GetActivityId()

File: service/history/historyEngine.go [hunk 7]
```diff
@@ -1996,323 +1957,6 @@ func (s *shardContextWrapper) NotifyNewHistoryEvent(event *historyEventNotificat
...
-
-	// Inherit decision task timeout from parent workflow execution if not provided on decision
-	if attributes.GetTaskStartToCloseTimeoutSeconds() <= 0 {
-		attributes.TaskStartToCloseTimeoutSeconds = common.Int32Ptr(parentInfo.DecisionTimeoutValue)
-	}
-
-	return nil
-}
-
 func validateStartWorkflowExecutionRequest(request *workflow.StartWorkflowExecutionRequest, maxIDLengthLimit int) error {
```
Comment (yux0) [inline service/history/historyEngine.go]: Can this also move to the new file?

Comment (wxing1292) [reply service/history/historyEngine.go]: this has nothing related to decision validation

File: service/history/decisionChecker_test.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+func (s *decisionAttrValidatorSuite) SetupTest() {
+	s.maxIDLengthLimit = 1000
+	s.validator = newDecisionAttrValidator(s.maxIDLengthLimit)
+}
+
+func (s *decisionAttrValidatorSuite) TearDownTest() {
+
+}
+
+func (s *decisionAttrValidatorSuite) TestValidateSignalExternalWorkflowExecutionAttributes() {
```
Comment (shreyassrivatsan) [inline service/history/decisionChecker_test.go]: We should move/add tests for the other cases here too.. 

If we need to add new tests, then that can be as a separate diff..

Comment (wxing1292) [reply service/history/decisionChecker_test.go]: this is pure refactoring + moving code

Comment (shreyassrivatsan) [reply service/history/decisionChecker_test.go]: Ok. Let's add a `TODO` to add tests for the other validations - and open an issue?

Comment (wxing1292) [reply service/history/decisionChecker_test.go]: #1948

**Pull request merged**